node/modules/storageminer.go

```diff
@@ -357,76 +356,187 @@ func SectorStorage(mctx helpers.MetricsCtx, lc fx.Lifecycle, lstor *paths.Local,
+			p.backOff(ctx)
+			continue
+		default:
+			log.Infow("Successfully acquired F3 participation lease.", "issuer", lease.Issuer, "expiry", lease.ValidityTerm)
+			p.previousTicket = ticket
+			return lease, true, nil
+		}
+	}
+	return api.F3ParticipationLease{}, false, ctx.Err()
+}
 
-			if ctx.Err() != nil {
-				return
-			}
-			if errors.Is(err, full.ErrF3Disabled) {
-				log.Errorf("Cannot participate in F3 as it is disabled: %+v", err)
-				return
-			}
-			if err != nil {
-				log.Errorf("while starting to participate in F3: %+v", err)
-				// use exponential backoff to avoid hotloop
-				timer.Reset(b.Duration())
-				select {
-				case <-ctx.Done():
-					return
-				case <-timer.C:
-				}
-				continue
-			}
-			if !ok {
-				log.Errorf("lotus node refused our lease, are you loadbalancing or did the miner just restart?")
-
-				sleepFor := b.Duration()
-				if d := time.Until(oldLease); d > 0 && d < sleepFor {
-					sleepFor = d
-				}
-				timer.Reset(sleepFor)
-				select {
-				case <-ctx.Done():
-					return
-				case <-timer.C:
-				}
-				continue
+func (p *f3Participator) awaitLeaseExpiry(ctx context.Context, lease api.F3ParticipationLease) error {
+	p.backoff.Reset()
+	for ctx.Err() == nil {
+		switch progress, err := p.node.F3GetProgress(ctx); {
+		case errors.Is(err, api.ErrF3Disabled):
+			log.Errorw("Cannot await F3 participation lease expiry as F3 is disabled.", "err", err)
+			return xerrors.Errorf("awaiting F3 participation lease expiry: %w", err)
+		case err != nil:
+			if p.backoff.Attempt() > float64(p.maxCheckProgressAttempts) {
+				log.Errorw("Too many failures while attempting to check F3  progress. Restarting participation.", "attempts", p.backoff.Attempt(), "err", err)
+				return nil
 			}
+			log.Errorw("Failed to check F3 progress while awaiting lease expiry. Retrying after backoff.", "attempts", p.backoff.Attempt(), "backoff", p.backoff.Duration(), "err", err)
+			p.backOff(ctx)
+		case progress.Instance+2 >= lease.ValidityTerm:
+			log.Infof("F3 progressed (%d) to within two instances of lease expiry (%d). Restarting participation.", progress.Instance, lease.ValidityTerm)
+			return nil
+		default:
+			remainingInstanceLease := lease.ValidityTerm - progress.Instance
+			log.Debugf("F3 participation lease is valid for further %d instances. Recking after %s.", remainingInstanceLease, p.checkProgressInterval)
+			p.backOffFor(ctx, p.checkProgressInterval)
+		}
+	}
+	return ctx.Err()
+}
 
-			// we have succeeded in giving a lease, reset the backoff
-			b.Reset()
+func (p *f3Participator) backOff(ctx context.Context) {
+	p.backOffFor(ctx, p.backoff.Duration())
+}
 
-			oldLease = newLease
-			// wait for the half of the lease time and then refresh
-			timer.Reset(leaseTime / 2)
-			select {
-			case <-ctx.Done():
-				return
-			case <-timer.C:
-			}
+func (p *f3Participator) backOffFor(ctx context.Context, d time.Duration) {
+	if !p.timer.Stop() {
+		<-p.timer.C
+	}
```

Comment:
This will deadlock if the timer already fired and we already read from the channel.

Reply:
I believe we just need to create a new timer each time (and don't store it anywhere).

Reply:
> This will deadlock if the timer already fired and we already read from the channel.

Despite the fact that backoff is never called from multiple goroutines?

> I believe we just need to create a new timer each time (and don't store it anywhere).

Will do, at the price of higher GC which i guess doesn't matter considering the backoff call frequency.

Reply:
> Despite the fact that backoff is never called from multiple goroutines?

 In this case, it would have deadlocked precisely the second time we call it because it would have expired the first time and we would have read from the channel. So `Stop()` would have returned false (nothing to stop) and reading from the channel would have blocked forever.